### ISSUE TITLE
Select multiple queues

### ISSUE THREAD

**Issue by thegreathir** [cameron314/readerwriterqueue#123]
What is the best way to wait behind multiple queues and being notified when a message is coming from any of them?
Actually I want something like select (in linux sockets or go programming language for channels).
What I am doing now is iterating over all of them and use the non-blocking api `try_dequeue`, but this solution is busy wait and not efficient.

**Comment by cameron314** [comment]
Short answer: You can't with the current API.

Longer answer: This could be done a few ways; perhaps the simplest is to use a semaphore, but this still requires iterating the queues on wakeup. (By sheer coincidence, this happens to exactly match the internal design of my [BlockingConcurrentQueue](https://github.com/cameron314/concurrentqueue/blob/master/blockingconcurrentqueue.h) -- enqueueing from multiple producers goes into separate internal sub-queues, and dequeueing waits on a shared semaphore.)

The best way to do it, as you say, is to use an OS-level primitive (select, poll, WaitForMultipleObjects) where the pool of objects waited upon is tied to corresponding queues. This is doable, but rather platform-specific.

**Comment by thegreathir** [comment]
Thanks

**Comment by stellarpower** [comment]
This would be a nice feature IMO.

Crystal makes heavy use of [select (as a language keyword)](https://crystal-lang.org/reference/1.20/syntax_and_semantics/select.html) which might give some ideas (good or bad). Language runtime-managed channels and "fibres" are their main threading primitive, taking inspiration from go. Although in my experience the way it was implemented left something to be desired, but when it comes to parallelism I do prefer to use normal threads and find a specific solution to my problem rather than force the problem into a given template (i.e. I use channels like this one when it fits a streaming or worker queue pattern and don't when it doesn't).

In particular I have found (in that, and in C++) that signalling an exit is something that some implementations miss or make more difficult. The listening end should block on either a new message or an exit request.

Right now I am thinking that possibly pulling the reading and writing, and the signalling, out of the class into a separate interface might be a way to go (or offering it as an option at least) may make this easy to work with. Then we could have various adapters and decouple the storage from the access. I think std::variant also offers a good interface we can use through visitors as lambdas.

So if we have a set of channels, very rough thoughts:
```C++

ReaderWriterQueue<int   > channelOne  (...);
ReaderWriterQueue<double> channelTwo  (...);
ReaderWriterQueue<Foo*  > channelThree(...);

Reader(channelOne, channelTwo, channelThree).loop(
    [&](int    nextValue){ ... },
    [&](double nextValue){ ... },
    [&](Foo*   nextValue){ ... }
);
// Exit handled gracefully by dropping out to here.


Reader(channelOne, channelTwo, channelThree).loopWithExitFunction(
    [&](int    nextValue)         { ... },
    [&](double nextValue)         { ... },
    [&](Foo*   nextValue)         { ... },
    [&](                ) -> bool {
        // Quit requested
        return true; // Acknowledged, break out of the loop; return false to cancel exit and keep looping
    }
);
```

Likewise, we could have a variadic equivalent of try_dequeue, where we consume all (or at most N)  available samples from all queues in the set, until they're empty, and then carry on our way.

Another option could be skills mixins, I don't know how much that might get in the way of the implementation and its efficiency, but it'd theoretically be nice to have a pretty basic master queue, and use a template specialisation that adds in or changes certain functionality (such as the ReaderWriter vs Concurrent queues, unbounded storage versus ring buffers). I've done this previously myself to good effect with Matrices - you have a master Matrix class and you can mix in through an enum template parameter things like dense/sparse storage, eager/lazy evaluation. It looks nicer to the user to have one master class, and also keeps a lid on the combinatorial explosion of options as multiple different classes tailored to each case. [NamedType](https://github.com/joboccara/NamedType) has IMO a good example of mixing in skills for strong typing.
